ts/src/provider.ts

```diff
@@ -195,39 +228,228 @@ export default class Provider {
    * Simulates the given transaction, returning emitted logs from execution.
    *
    * @param tx      The transaction to send.
-   * @param signers The set of signers in addition to the provdier wallet that
-   *                will sign the transaction.
+   * @param signers The signers of the transaction.
    * @param opts    Transaction confirmation options.
    */
   async simulate(
     tx: Transaction,
-    signers?: Array<Signer | undefined>,
-    opts: ConfirmOptions = this.opts
-  ): Promise<RpcResponseAndContext<SimulatedTransactionResponse>> {
-    if (signers === undefined) {
-      signers = [];
-    }
-
+    signers?: Signer[],
+    commitment?: Commitment,
+    includeAccounts?: boolean | PublicKey[]
+  ): Promise<SuccessfulTxSimulationResponse> {
     tx.feePayer = this.wallet.publicKey;
-    tx.recentBlockhash = (
-      await this.connection.getRecentBlockhash(
-        opts.preflightCommitment ?? this.opts.preflightCommitment
-      )
-    ).blockhash;
 
-    tx = await this.wallet.signTransaction(tx);
-    signers
-      .filter((s): s is Signer => s !== undefined)
-      .forEach((kp) => {
-        tx.partialSign(kp);
-      });
-
-    return await simulateTransaction(
+    const result = await simulateTransaction(
       this.connection,
       tx,
-      opts.commitment ?? this.opts.commitment ?? "processed"
+      this.wallet,
+      signers,
+      commitment,
+      includeAccounts
+    );
+
+    if (result.value.err) {
+      throw new SimulateError(result.value);
+    }
+
+    return result.value;
+  }
+}
+
+// copy from @solana/web3.js that allows a wallet to sign
+// and has a commitment param
+async function simulateTransaction(
+  connection: Connection,
+  transactionOrMessage: Transaction,
+  wallet?: Wallet,
+  signers?: Array<Signer>,
+  commitment?: Commitment,
+  includeAccounts?: boolean | Array<PublicKey>
+): Promise<RpcResponseAndContext<SimulatedTransactionResponse>> {
+  let transaction;
+  if (transactionOrMessage instanceof Transaction) {
+    transaction = transactionOrMessage;
+  } else {
+    transaction = Transaction.populate(transactionOrMessage);
+  }
+
+  if (transaction.nonceInfo && signers) {
+    if (wallet) {
+      transaction = await wallet.signTransaction(transaction);
```

Comment:
We should remove the wallet parameter and force the caller to sign prior to passing in the tx.

Reply:
I remember why I did it this way now. `simulateTransaction` has all this logic to efficiently get the blockhash which and that has to be done *before* the first signature.